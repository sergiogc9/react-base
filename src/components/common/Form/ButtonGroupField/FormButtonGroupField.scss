@import "@src/scss/_globals.scss";
@import "@src/scss/_mixins.scss";
@import "@src/scss/_variables.scss";

.form-button-group-field-wrapper {
	.form-button-group-label{
		font-size: 1.2rem;
	}

	.form-button-group-field-buttons {
		margin-top: 1.25rem;

		.form-button-group-field-button {
			font-size: 1.2rem;
			text-transform: none;
			padding: 0.8rem 1.6rem;

			&.selected {
				background-color: $launchmetrics-green;
				border: 1px solid $launchmetrics-green;
			}

			&:not(.selected){
				border: 1px solid $md-grey-300;
			}

			@include respond(tab-land) {
				height: 3.2rem;
			}
		}
	}
}
