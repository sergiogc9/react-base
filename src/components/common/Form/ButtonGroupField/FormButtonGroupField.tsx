import React from 'react';
import { Button } from 'react-md';
import find from 'lodash/find';

import { ButtonGroupFieldElement } from '@src/types/form';
import { formFieldBaseValidation } from '@src/components/common/Form/Form';
import { useForceFieldValue } from '@src/lib/hooks';

import './FormButtonGroupField.scss';

export type ComponentProps = {
	element: ButtonGroupFieldElement,
	forceValue?: any,
	onChangeOption: (elementId: string, value: string, isValid: boolean) => void
};

const __validateButtonGroupField = (element: ButtonGroupFieldElement, newValue: any) => {
	return formFieldBaseValidation(element, newValue);
};

const FormButtonGroupField = (props: ComponentProps) => {
	const { element, forceValue } = props;
	const { onChangeOption } = props;

	const defaultOption = element.defaultValue ? find(element.options, { value: element.defaultValue }) : null;

	const [value, setValue] = React.useState(defaultOption ? defaultOption.value : element.options[0].value);
	useForceFieldValue(forceValue, null, element, setValue, onChangeOption, __validateButtonGroupField);

	const clickHandler = React.useCallback((event: React.MouseEvent<HTMLElement>) => {
		const id = event.currentTarget.id.replace(element.id + '-', '');
		const option = find(element.options, { id });
		setValue(option!.value);
		onChangeOption(element.id, option!.value, __validateButtonGroupField(element, option!.value));
	}, [element, onChangeOption]);

	return (
		<div id={`formButtonGroupField-${element.id}`} className={`form-button-group-field-wrapper ${element.className || ''}`}>
			{element.label ? <span className="form-button-group-label">{element.label}</span> : null}
			<div className="form-button-group-field-buttons">
				{element.options.map(choice =>
					<Button
						id={`${element.id}-${choice.id}`}
						key={`form-button-group-${choice.id}`}
						flat
						className={`form-button-group-field-button ${value === choice.value ? 'selected' : ''}`}
						onClick={clickHandler}
						{...choice.buttonProps}
					>
						{choice.label}
					</Button>
				)}
			</div>
		</div>
	);
};

export default React.memo(FormButtonGroupField);
