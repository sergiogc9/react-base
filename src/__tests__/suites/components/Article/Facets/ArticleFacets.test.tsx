import { shallow } from "enzyme";
import moment from 'moment';
import React from "react";

import ArticleFacets from "@src/components/Article/Facets/ArticleFacets";
import GA from '@src/lib/googleAnalytics';
import FormButtonGroupField from '@src/components/common/Form/ButtonGroupField/FormButtonGroupField';

let onToggleDrawer: any;
let onSetPeriod: any;
let onSetDateType: any;

const beginDate = moment().startOf('day').subtract(1, 'week').toDate();
const endDate = moment().endOf('day').toDate();

describe("<ArticleFacets />", () => {
	let wrapper: any;

	beforeAll(() => {
		window.matchMedia = jest.fn((query: string) => ({
			matches: query.indexOf('(min-width: 1025px)') !== -1
		}));
	});

	beforeEach(() => {
		onToggleDrawer = jest.fn();
		onSetPeriod = jest.fn();
		onSetDateType = jest.fn();
		wrapper = shallow(<ArticleFacets
			drawerVisible={true}
			searchPeriod="last_month"
			dateType='publication_date'
			onToggleDrawer={onToggleDrawer}
			onSetPeriod={onSetPeriod}
			onSetDateType={onSetDateType}
		/>);
	});

	it("Test component snapshot", () => {
		expect(wrapper.debug()).toMatchSnapshot();
		expect(onToggleDrawer).toHaveBeenCalled();
	});

	it("Test with custom date ranges", () => {
		wrapper.setProps({ searchPeriod: "custom", beginDate, endDate });
		expect(wrapper.debug()).toMatchSnapshot();
	});

	it("Test period select change", () => {
		const trackEvent = jest.fn();
		GA.trackEvent = trackEvent;
		expect(trackEvent).toHaveBeenCalledTimes(0);
		expect(onSetPeriod).toHaveBeenCalledTimes(0);
		wrapper.find('#searchFacetsPeriodSelect').at(0).props().onChange("last_week");
		expect(trackEvent).toHaveBeenCalledTimes(1);
		expect(onSetPeriod).toHaveBeenCalledTimes(1);
		wrapper.find('#searchFacetsPeriodSelect').at(0).props().onChange("custom");
		expect(trackEvent).toHaveBeenCalledTimes(2);
		expect(onSetPeriod).toHaveBeenCalledTimes(2);
	});

	it("Component onChange handler", () => {
		wrapper.find(FormButtonGroupField).at(0).props().onChangeOption('test', 'integration_date', true);
		expect(onSetDateType).toHaveBeenCalledTimes(1);
	});
});
