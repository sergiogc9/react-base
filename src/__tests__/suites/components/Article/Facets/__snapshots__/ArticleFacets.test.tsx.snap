// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArticleFacets /> Test component snapshot 1`] = `
"<Drawer id=\\"facetsDrawer\\" type=\\"floating\\" inline={true} visible={true} clickableDesktopOverlay={false} overlay={false} onVisibilityChange={[Function: onVisibilityChange]} onMediaTypeChange={[Function: onMediaTypeChange]} defaultMedia=\\"mobile\\" mobileType=\\"temporary\\" mobileMinWidth={320} tabletType=\\"persistent\\" tabletMinWidth={768} desktopType=\\"full-height\\" desktopMinWidth={1025} position=\\"left\\" transitionDuration={300} autoclose={true} constantType={true}>
  <div className=\\"header\\">
    <div className=\\"title\\">
      results.sidebar.title
    </div>
    <div className=\\"description\\">
      results.sidebar.description_1
    </div>
    <div className=\\"description\\">
      results.sidebar.description_2
    </div>
  </div>
  <SelectField id=\\"searchFacetsPeriodSelect\\" label=\\"filters.period.value\\" placeholder=\\"filters.period.value\\" menuItems={{...}} defaultValue=\\"last_month\\" value=\\"last_month\\" position=\\"tr\\" onChange={[Function: onChange]} anchor={{...}} fixedTo={{...}} itemLabel=\\"label\\" itemValue=\\"value\\" itemProps=\\"getProps\\" dropdownIcon={{...}} lineDirection=\\"left\\" defaultVisible={false} keyboardMatchingTimeout={1000} transitionName=\\"md-drop\\" transitionTime={300} repositionOnScroll={true} repositionOnResize={false} />
  <Memo(FormButtonGroupField) element={{...}} onChangeOption={[Function]} forceValue=\\"publication_date\\" />
  <Connect(SearchFacets) />
</Drawer>"
`;

exports[`<ArticleFacets /> Test with custom date ranges 1`] = `
"<Drawer id=\\"facetsDrawer\\" type=\\"floating\\" inline={true} visible={true} clickableDesktopOverlay={false} overlay={false} onVisibilityChange={[Function: onVisibilityChange]} onMediaTypeChange={[Function: onMediaTypeChange]} defaultMedia=\\"mobile\\" mobileType=\\"temporary\\" mobileMinWidth={320} tabletType=\\"persistent\\" tabletMinWidth={768} desktopType=\\"full-height\\" desktopMinWidth={1025} position=\\"left\\" transitionDuration={300} autoclose={true} constantType={true}>
  <div className=\\"header\\">
    <div className=\\"title\\">
      results.sidebar.title
    </div>
    <div className=\\"description\\">
      results.sidebar.description_1
    </div>
    <div className=\\"description\\">
      results.sidebar.description_2
    </div>
  </div>
  <SelectField id=\\"searchFacetsPeriodSelect\\" label=\\"filters.period.value\\" placeholder=\\"filters.period.value\\" menuItems={{...}} defaultValue=\\"custom\\" value=\\"custom\\" position=\\"tr\\" onChange={[Function: onChange]} anchor={{...}} fixedTo={{...}} itemLabel=\\"label\\" itemValue=\\"value\\" itemProps=\\"getProps\\" dropdownIcon={{...}} lineDirection=\\"left\\" defaultVisible={false} keyboardMatchingTimeout={1000} transitionName=\\"md-drop\\" transitionTime={300} repositionOnScroll={true} repositionOnResize={false} />
  <div className=\\"dateRanges\\">
     
    <Connect(SearchDateRange) />
  </div>
  <Memo(FormButtonGroupField) element={{...}} onChangeOption={[Function]} forceValue=\\"publication_date\\" />
  <Connect(SearchFacets) />
</Drawer>"
`;
